cli/internal/run/global_hash.go

```diff
@@ -103,6 +107,11 @@ func calculateGlobalHash(
 		return GlobalHashable{}, err
 	}
 
+	// The only way we can add env vars into the hash via matching is via THASH.
+	if !isStructuredOutput && len(globalHashableEnvVars.BySource.Matching) > 0 {
+		ui.Warn(fmt.Sprintf("[DEPRECATED] Using .*THASH.* to specify an environment variable for inclusion into the hash is deprecated. You specified: %s.", strings.Join(globalHashableEnvVars.BySource.Matching.Names(), ", ")))
+	}
```

Comment:
I intentionally elected to do minimally-invasive changes here since I want to remove this in 1.10 (/whenever we remove the `experimental` flag on strict env). All of the semi-questionable code here gets deleted in that future PR and optimizing the implementation for insta-delete code creates unused abstractions that linger.

As for structured vs. non-structured output: in Rust-land we should make a _much_ better logging story where the thing that handles writing to stdout knows if we're spitting out structured output. Right now that's just a global toggle over there too:

https://github.com/vercel/turbo/blob/df0a60ec028b486f3865d831046d0918de469ac2/crates/turborepo-lib/src/shim.rs#L29-L39